### DISCUSSION TITLE
FFMPEG vs YouTube-DL

### DISCUSSION THREAD

**Discussion by iim-ayush** [discordjs/voice#134] [General]
I did a research again on this topic.
FFMPEG codes [here](https://github.com/discordjs/voice/discussions/123) second one. [With FFMPEG reconnecting arguments]
YouTube-DL codes [here](https://github.com/discordjs/voice/tree/main/examples/music-bot) Native Voice Example (music-bot).

**Note** : Example doesn't have `inLineVolume` enabled by default, so I have turned that ON for both codes.

**Procedure of conducting the test** :
1. Run these codes.
2. Now wait 10-15 seconds to pass and then record the memory usages. 

First `FFMPEG` code, 
Results [here](https://cdn.discordapp.com/attachments/855114956146868244/855114978033401927/ffmpeg-test-1.png) or find it down [below](https://github.com/discordjs/voice/discussions/134#discussioncomment-884264)

As we can see, after running bot in 1 guild, total resources that were used were 55 MB(approx.).
FFMPEG (using 2.7 MB) is used to create a volume interface for bot [IGNORE This one in all cases]
Real FFMPEG used 6.6 MB to play song.

As we connect bot in 2nd guild, total resources that were used were 80 MB (approx.).
Again FFMPEG used 6.8 MB to play song.

Per guild increase in resources(including all (creating audio player, audio resource, creating a FFMPEG instance etc.)) = 80 - 55 MB = **25 MB**

Now moving on to `YouTube-DL` code, 
Results [here](https://cdn.discordapp.com/attachments/855114956146868244/855114998668722196/youtube-dl-test-2.png) or find it down [below](https://github.com/discordjs/voice/discussions/134#discussioncomment-884264)

As we can see, after running bot in 1 guild, total resources that were used 91 MB [Which is 10 MB more than running 2 guilds with FFMPEG]
YouTube-DL used 27 MB (compared to 6.8 MB for ffmpeg ) to play same song.

As we connect bot in 2nd guild, total resources that were used were 140 MB (approx.).
Again YouTube-DL used 27 MB (approx.) to play song

Per guild increase in resources (including all (creating audio player, audio resource, creating a YouTube-DL instance etc.)) = 140 - 91 MB = **49 MB**

Now let's compare features of both:
1. Streaming song
2. Powers of connecting back if disconnected for any reasons.

For me personally, I will always prefer FFMPEG over YouTube-DL due to better memory management.
I want to know your thoughts. So let me know by commenting down below.

**Comment by iim-ayush** [comment]
### FFMPEG Result
![ffmpeg-test-1](https://user-images.githubusercontent.com/65385476/122437191-19939800-cfb7-11eb-9cc3-7776340e3ef2.png)
------------------------------------------------------------------------------------------------------------------------------------------------------
###  YouTube-DL Result
![youtube-dl-test-2](https://user-images.githubusercontent.com/65385476/122437248-2a440e00-cfb7-11eb-80dd-69f206422fab.png)

**Reply by iCrawl** [reply]
Thats a really bad way to measure performance and CPU/Memory usage.

Not to mention how Windows will not be your target system when that bot goes live to run on a server.

**Reply by iim-ayush** [reply]
> Thats a really bad way to measure performance and CPU/Memory usage.
> 
> Not to mention how Windows will not be your target system when that bot goes live to run on a server.

So tell me one, I can clearly show those results there also.

**Reply by iim-ayush** [reply]
> Thats a really bad way to measure performance and CPU/Memory usage.
> 
> Not to mention how Windows will not be your target system when that bot goes live to run on a server.

Then show me your results in your good way.

**Comment by amishshah** [comment]
I think it's also worth taking into account that the examples with `youtube-dl` have lower CPU utilisation compared to the version of the music bot that is running with FFmpeg by your own benchmarks.

It's definitely not wrong to use FFmpeg, but for our own music bot which doesn't use any extra filters, it makes sense to go with youtube-dl because:

- youtube-dl supports audio from many more sources than just YouTube
- we know that youtube-dl will work for reliably downloading/streaming videos - it's what it was built for
- code is clearer - no need for extra FFmpeg arguments
- no extra CPU time wasted on transcoding audio stream for no reason.

At the end of the day, use whatever works best on your own system.
